lucille-core/src/main/java/com/kmwllc/lucille/core/Runner.java

```diff
@@ -210,49 +210,59 @@ public static RunResult run(Config config, RunType type) throws Exception {
-        publisherMMFactory = PublisherMessageManagerFactory.getKafkaFactory(config);
-      }
+    try {
+      List<ConnectorResult> connectorResults = new ArrayList<>();
+
+      boolean startWorkerAndIndexer = !type.equals(RunType.KAFKA_DISTRIBUTED);
+      boolean bypassSolr = type.equals(RunType.TEST);
+
+      Map<String, PersistingLocalMessageManager> history = type.equals(RunType.TEST) ? new HashMap<>() : null;
+
+      for (Connector connector : connectors) {
+
+        WorkerMessageManagerFactory workerMMFactory;
+        IndexerMessageManagerFactory indexerMMFactory;
+        PublisherMessageManagerFactory publisherMMFactory;
+
+        if (RunType.TEST.equals(type)) {
+          PersistingLocalMessageManager manager = new PersistingLocalMessageManager();
+          history.put(connector.getName(), manager);
+          workerMMFactory = WorkerMessageManagerFactory.getConstantFactory(manager);
+          indexerMMFactory = IndexerMessageManagerFactory.getConstantFactory(manager);
+          publisherMMFactory = PublisherMessageManagerFactory.getConstantFactory(manager);
+        } else if (RunType.LOCAL.equals(type)) {
+          LocalMessageManager manager = new LocalMessageManager(config);
+          workerMMFactory = WorkerMessageManagerFactory.getConstantFactory(manager);
+          indexerMMFactory = IndexerMessageManagerFactory.getConstantFactory(manager);
+          publisherMMFactory = PublisherMessageManagerFactory.getConstantFactory(manager);
+        } else { // RunType.KAFKA_LOCAL.equals(type) || RunType.KAFKA_DISTRIBUTED.equals(type)
+          workerMMFactory = WorkerMessageManagerFactory.getKafkaFactory(config, connector.getPipelineName());
+          indexerMMFactory = IndexerMessageManagerFactory.getKafkaFactory(config, connector.getPipelineName());
+          publisherMMFactory = PublisherMessageManagerFactory.getKafkaFactory(config);
+        }
 
-      ConnectorResult result =
-        runConnectorWithComponents(config, runId, connector,
-          workerMMFactory, indexerMMFactory, publisherMMFactory, startWorkerAndIndexer, bypassSolr);
+        ConnectorResult result =
+          runConnectorWithComponents(config, runId, connector,
+            workerMMFactory, indexerMMFactory, publisherMMFactory, startWorkerAndIndexer, bypassSolr);
 
-      connectorResults.add(result);
+        connectorResults.add(result);
 
-      if (!result.getStatus()) {
-        log.error("Aborting run because " + connector.getName() + " failed.");
-        return new RunResult(false, connectors, connectorResults, history, runId);
+        if (!result.getStatus()) {
+          log.error("Aborting run because " + connector.getName() + " failed.");
+          return new RunResult(false, connectors, connectorResults, history, runId);
+        }
+      }
+      return new RunResult(true, connectors, connectorResults, history, runId);
+    } finally {
+      // make sure none of the connectors linger beyond this run.
```

Comment:
Do we understand how connectors were lingering before? runConnectorWithComponents() calls runConnector() which is set up as a try/finally with connector.close() in the finally. If that code isn't ensuring that connectors are closed, I'd like to know where the leak is happening...

Reply:
The SolrTestFramework is checking for lingering threads/objects. It is detecting that some solr clients that were created by these tests were not released.

`[InternalHttpClient]
2023-08-15T19:19:19.8879350Z org.apache.solr.common.util.ObjectReleaseTracker$ObjectTrackerException: org.apache.http.impl.client.InternalHttpClient
2023-08-15T19:19:19.8880460Z 	at org.apache.solr.common.util.ObjectReleaseTracker.track(ObjectReleaseTracker.java:42)
2023-08-15T19:19:19.8881810Z 	at org.apache.solr.client.solrj.impl.HttpClientUtil.createClient(HttpClientUtil.java:341)
2023-08-15T19:19:19.8882760Z 	at org.apache.solr.client.solrj.impl.HttpClientUtil.createClient(HttpClientUtil.java:350)
2023-08-15T19:19:19.8883760Z 	at org.apache.solr.client.solrj.impl.HttpClientUtil.createClient(HttpClientUtil.java:288)
2023-08-15T19:19:19.8884960Z 	at org.apache.solr.client.solrj.impl.HttpClientUtil.createClient(HttpClientUtil.java:275)
2023-08-15T19:19:19.8886040Z 	at org.apache.solr.client.solrj.impl.HttpSolrClient.<init>(HttpSolrClient.java:204)
2023-08-15T19:19:19.8886860Z 	at org.apache.solr.client.solrj.impl.HttpSolrClient$Builder.build(HttpSolrClient.java:968)
2023-08-15T19:19:19.8887860Z 	at com.kmwllc.lucille.util.SolrUtils.getSolrClient(SolrUtils.java:52)
2023-08-15T19:19:19.8888830Z 	at com.kmwllc.lucille.indexer.SolrIndexer.getSolrClient(SolrIndexer.java:45)
2023-08-15T19:19:19.8889720Z 	at com.kmwllc.lucille.indexer.SolrIndexer.<init>(SolrIndexer.java:41)
2023-08-15T19:19:19.8890760Z 	at com.kmwllc.lucille.indexer.IndexerFactory.fromConfig(IndexerFactory.java:42)
2023-08-15T19:19:19.8891710Z 	at com.kmwllc.lucille.core.Runner.runConnectorWithComponents(Runner.java:395)
2023-08-15T19:19:19.8892500Z 	at com.kmwllc.lucille.core.Runner.run(Runner.java:244)
2023-08-15T19:19:19.8893390Z 	at com.kmwllc.lucille.core.RunnerTest.testIndexerConnectFailure(RunnerTest.java:734)
2023-08-15T19:19:19.8894210Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-08-15T19:19:19.8895940Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
2023-08-15T19:19:19.8897120Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-08-15T19:19:19.8898620Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
2023-08-15T19:19:19.8899480Z 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
2023-08-15T19:19:19.8900440Z 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2023-08-15T19:19:19.8901610Z 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
2023-08-15T19:19:19.8902560Z 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2023-08-15T19:19:19.8903030Z 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
2023-08-15T19:19:19.8903810Z 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
2023-08-15T19:19:19.8905180Z 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
2023-08-15T19:19:19.8906230Z 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
2023-08-15T19:19:19.8911140Z 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
2023-08-15T19:19:19.8912130Z 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
2023-08-15T19:19:19.8912890Z 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
2023-08-15T19:19:19.8913940Z 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
2023-08-15T19:19:19.8914930Z 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
2023-08-15T19:19:19.8915830Z 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
2023-08-15T19:19:19.8916750Z 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
2023-08-15T19:19:19.8917820Z 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
2023-08-15T19:19:19.8918780Z 	at org.junit.runners.Suite.runChild(Suite.java:128)
2023-08-15T19:19:19.8919600Z 	at org.junit.runners.Suite.runChild(Suite.java:27)
2023-08-15T19:19:19.8920820Z 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
2023-08-15T19:19:19.8922340Z 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
2023-08-15T19:19:19.8923240Z 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
2023-08-15T19:19:19.8924260Z 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
2023-08-15T19:19:19.8925260Z 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
2023-08-15T19:19:19.8926010Z 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
2023-08-15T19:19:19.8926890Z 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
2023-08-15T19:19:19.8927950Z 	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
2023-08-15T19:19:19.8928990Z 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
2023-08-15T19:19:19.8930210Z 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
2023-08-15T19:19:19.8934660Z 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
2023-08-15T19:19:19.8935920Z 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
2023-08-15T19:19:19.8937040Z 	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)
2023-08-15T19:19:19.8938110Z 	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:383)
2023-08-15T19:19:19.8939370Z 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:344)
2023-08-15T19:19:19.8954310Z 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
2023-08-15T19:19:19.8961480Z 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:417)
2023-08-15T19:19:19.8962940Z 
2023-08-15T19:19:19.8979730Z  expected null, but was:<ObjectTracker found 1 object(s) that were not released!!! [InternalHttpClient]
2023-08-15T19:19:19.8981080Z org.apache.solr.common.util.ObjectReleaseTracker$ObjectTrackerException: org.apache.http.impl.client.InternalHttpClient
2023-08-15T19:19:19.8983100Z 	at org.apache.solr.common.util.ObjectReleaseTracker.track(ObjectReleaseTracker.java:42)
2023-08-15T19:19:19.8984310Z 	at org.apache.solr.client.solrj.impl.HttpClientUtil.createClient(HttpClientUtil.java:341)
2023-08-15T19:19:19.8990080Z 	at org.apache.solr.client.solrj.impl.HttpClientUtil.createClient(HttpClientUtil.java:350)
2023-08-15T19:19:19.8991040Z 	at org.apache.solr.client.solrj.impl.HttpClientUtil.createClient(HttpClientUtil.java:288)
2023-08-15T19:19:19.8992030Z 	at org.apache.solr.client.solrj.impl.HttpClientUtil.createClient(HttpClientUtil.java:275)
2023-08-15T19:19:19.8992910Z 	at org.apache.solr.client.solrj.impl.HttpSolrClient.<init>(HttpSolrClient.java:204)
2023-08-15T19:19:19.8993850Z 	at org.apache.solr.client.solrj.impl.HttpSolrClient$Builder.build(HttpSolrClient.java:968)
2023-08-15T19:19:19.8994870Z 	at com.kmwllc.lucille.util.SolrUtils.getSolrClient(SolrUtils.java:52)
2023-08-15T19:19:19.8995850Z 	at com.kmwllc.lucille.indexer.SolrIndexer.getSolrClient(SolrIndexer.java:45)
2023-08-15T19:19:19.8996720Z 	at com.kmwllc.lucille.indexer.SolrIndexer.<init>(SolrIndexer.java:41)
2023-08-15T19:19:19.9008480Z 	at com.kmwllc.lucille.indexer.IndexerFactory.fromConfig(IndexerFactory.java:42)
2023-08-15T19:19:19.9009920Z 	at com.kmwllc.lucille.core.Runner.runConnectorWithComponents(Runner.java:395)
2023-08-15T19:19:19.9010370Z 	at com.kmwllc.lucille.core.Runner.run(Runner.java:244)
2023-08-15T19:19:19.9010830Z 	at com.kmwllc.lucille.core.RunnerTest.testIndexerConnectFailure(RunnerTest.java:734)
2023-08-15T19:19:19.9011300Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-08-15T19:19:19.9011800Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
2023-08-15T19:19:19.9012370Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-08-15T19:19:19.9013280Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
2023-08-15T19:19:19.9013970Z 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
2023-08-15T19:19:19.9014440Z 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2023-08-15T19:19:19.9014930Z 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
2023-08-15T19:19:19.9015420Z 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2023-08-15T19:19:19.9015840Z 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
2023-08-15T19:19:19.9016240Z 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
2023-08-15T19:19:19.9016640Z 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
2023-08-15T19:19:19.9017060Z 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
2023-08-15T19:19:19.9017530Z 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
2023-08-15T19:19:19.9017940Z 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
2023-08-15T19:19:19.9018280Z 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
2023-08-15T19:19:19.9018660Z 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
2023-08-15T19:19:19.9019030Z 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
2023-08-15T19:19:19.9019430Z 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
2023-08-15T19:19:19.9019790Z 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
2023-08-15T19:19:19.9020130Z 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
2023-08-15T19:19:19.9020460Z 	at org.junit.runners.Suite.runChild(Suite.java:128)
2023-08-15T19:19:19.9021090Z 	at org.junit.runners.Suite.runChild(Suite.java:27)
2023-08-15T19:19:19.9021420Z 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
2023-08-15T19:19:19.9021780Z 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
2023-08-15T19:19:19.9022440Z 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
2023-08-15T19:19:19.9022810Z 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
2023-08-15T19:19:19.9023170Z 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
2023-08-15T19:19:19.9023530Z 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
2023-08-15T19:19:19.9026100Z 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
2023-08-15T19:19:19.9026500Z 	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
2023-08-15T19:19:19.9027030Z 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
2023-08-15T19:19:19.9027610Z 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
2023-08-15T19:19:19.9028140Z 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
2023-08-15T19:19:19.9028660Z 	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
2023-08-15T19:19:19.9029190Z 	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)
2023-08-15T19:19:19.9029760Z 	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:383)
2023-08-15T19:19:19.9030320Z 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:344)
2023-08-15T19:19:19.9030810Z 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
2023-08-15T19:19:19.9031260Z 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:417)
2023-08-15T19:19:19.9031500Z `

Reply:
Actually I think it's just the Indexer that is leaking. The indexer is never closed the scenario when indexer.validateConnection fails because the exception is thrown before the Indexer is run in a thread.

```       
 if (!indexer.validateConnection()) {
    String msg = "Indexer could not connect.";
    log.error(msg);
    return new ConnectorResult(connector, publisher, false, msg);
  }

  indexerThread = new Thread(indexer);
  indexerThread.start();```

Reply:
Agree -- was just going to write the same thing. I think we can & should fix this by calling indexer.closeConnection() in that if block that you posted. I don't think we should add the the large try/finally in Runner.run because there's other code that's aiming to make sure connectors are always closed & if that code is not working we should fix it at the source.

Reply:
I reverted the previous changes to Runner.run and added the indexer.closeConnection().